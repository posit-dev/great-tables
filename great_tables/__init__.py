# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from . import loc, style, vals
from ._helpers import (
    LETTERS,
    define_units,
    google_font,
    html,
    letters,
    md,
    nanoplot_options,
    pct,
    px,
    random_id,
    system_fonts,
)
from ._styles import FromColumn as from_column
from .gt import GT

__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
