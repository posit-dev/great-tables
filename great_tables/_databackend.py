from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
