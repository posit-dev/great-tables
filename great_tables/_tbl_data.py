from __future__ import annotations

import warnings

from functools import singledispatch
from typing import Any, Dict, List, Union, Callable, Tuple, TYPE_CHECKING
from typing_extensions import TypeAlias

from ._databackend import AbstractBackend


# Define databackend types ----
# These are resolved lazily (e.g. on isinstance checks) when run dynamically,
# or imported directly during type checking.

if TYPE_CHECKING:
    import pandas as pd
    import polars as pl
    import numpy as np

    # the class behind selectors
    from polars.selectors import _selector_proxy_

    PdDataFrame = pd.DataFrame
    PlDataFrame = pl.DataFrame
    PlSelectExpr = _selector_proxy_
    PlExpr = pl.Expr

    PdSeries = pd.Series
    PlSeries = pl.Series

    PdNA = pd.NA
    PlNull = pl.Null

    NpNan = np.nan

    DataFrameLike = Union[PdDataFrame, PlDataFrame]
    SeriesLike = Union[PdSeries, PlSeries]
    TblData = DataFrameLike

else:
    from abc import ABC

    # we just need this as a static type hint, but singledispatch tries to resolve
    # any hints at runtime. So we need some value for it.
    from typing import Any as _selector_proxy_

    class PdDataFrame(AbstractBackend):
        _backends = [("pandas", "DataFrame")]

    class PlDataFrame(AbstractBackend):
        _backends = [("polars", "DataFrame")]

    class PlSelectExpr(AbstractBackend):
        _backends = [("polars.selectors", "_selector_proxy_")]

    class PlExpr(AbstractBackend):
        _backends = [("polars", "Expr")]

    class PdSeries(AbstractBackend):
        _backends = [("pandas", "Series")]

    class PlSeries(AbstractBackend):
        _backends = [("polars", "Series")]

    class PdNA(AbstractBackend):
        _backends = [("pandas", "NA")]

    class PlNull(AbstractBackend):
        _backends = [("polars", "Null")]

    class NpNan(AbstractBackend):
        _backends = [("numpy", "nan")]

    # TODO: these types are imported throughout gt, so we need to either put
    # those imports under TYPE_CHECKING, or continue to make available dynamically here.
    class DataFrameLike(ABC):
        """Represent some DataFrame"""

    class SeriesLike(ABC):
        """Represent some Series"""

    DataFrameLike.register(PdDataFrame)
    DataFrameLike.register(PlDataFrame)
    SeriesLike.register(PdSeries)
    SeriesLike.register(PlSeries)

    TblData = DataFrameLike


# utils ----


def _raise_not_implemented(data):
    raise NotImplementedError(f"Unsupported data type: {type(data)}")


# generic functions ----


# copy_data ----
@singledispatch
def copy_data(data: DataFrameLike) -> DataFrameLike:
    """Copy the stored table data"""
    _raise_not_implemented(data)


@copy_data.register(PdDataFrame)
def _(data: PdDataFrame):
    return data.copy()


@copy_data.register(PlDataFrame)
def _(data: PlDataFrame):
    return data.clone()


# get_column_names ----
@singledispatch
def get_column_names(data: DataFrameLike) -> List[str]:
    """Get a list of column names from the input data table"""
    _raise_not_implemented(data)


@get_column_names.register(PdDataFrame)
def _(data: PdDataFrame):
    return data.columns.tolist()


@get_column_names.register(PlDataFrame)
def _(data: PlDataFrame):
    return data.columns


# n_rows ----


@singledispatch
def n_rows(data: DataFrameLike) -> int:
    """Get the number of rows from the input data table"""
    raise _raise_not_implemented(data)


@n_rows.register(PdDataFrame)
@n_rows.register(PlDataFrame)
def _(data):
    return len(data)


# _get_cell ----


@singledispatch
def _get_cell(data: DataFrameLike, row: int, column: str) -> Any:
    """Get the content from a single cell in the input data table"""

    _raise_not_implemented(data)


@_get_cell.register(PlDataFrame)
def _(data, row: int, column: str):
    return data[column][row]


@_get_cell.register(PdDataFrame)
def _(data, row, col):
    col_ii = data.columns.get_loc(col)

    if not isinstance(col_ii, int):
        raise ValueError("Column named " + col + " matches multiple columns.")

    return data.iloc[row, col_ii]


# _set_cell ----


@singledispatch
def _set_cell(data: DataFrameLike, row: int, column: str, value: Any):
    _raise_not_implemented(data)


@_set_cell.register(PdDataFrame)
def _(data, row: int, column: str, value: Any):
    # TODO: This assumes column names are unique
    # if this is violated, get_loc will return a mask
    col_indx = data.columns.get_loc(column)
    data.iloc[row, col_indx] = value


@_set_cell.register(PlDataFrame)
def _(data, row: int, column: str, value: Any):
    data[row, column] = value


# _get_column_dtype ----


@singledispatch
def _get_column_dtype(data: DataFrameLike, column: str) -> str:
    """Get the data type for a single column in the input data table"""
    return data[column].dtype


# reorder ----


@singledispatch
def reorder(data: DataFrameLike, rows: List[int], columns: List[str]) -> DataFrameLike:
    """Return a re-ordered DataFrame."""
    _raise_not_implemented(data)


@reorder.register
def _(data: PdDataFrame, rows: List[int], columns: List[str]) -> PdDataFrame:
    # note that because loc is label based, we need
    # reset index to allow us to use integer indexing on the rows
    # note that this means the index is not preserved when reordering pandas
    return data.iloc[rows, :].loc[:, columns]


@reorder.register
def _(data: PlDataFrame, rows: List[int], columns: List[str]) -> PlDataFrame:
    return data[rows, columns]


# group_splits ----
@singledispatch
def group_splits(data: DataFrameLike, group_key: str) -> Dict[Any, List[int]]:
    raise NotImplementedError(f"Unsupported data type: {type(data)}")


@group_splits.register
def _(data: PdDataFrame, group_key: str) -> Dict[Any, List[int]]:
    g_df = data.groupby(group_key)
    return {k: list(v) for k, v in g_df.indices.items()}


@group_splits.register
def _(data: PlDataFrame, group_key: str) -> Dict[Any, List[int]]:
    # TODO: should ensure row count name isn't already in data
    import polars as pl

    # with_row_index supercedes with_row_count
    meth_row_number = getattr(data, "with_row_index", None)
    if not meth_row_number:
        meth_row_number = data.with_row_count

    groups = meth_row_number("__row_count__").group_by(group_key).agg(pl.col("__row_count__"))

    res = dict(zip(groups[group_key].to_list(), groups["__row_count__"].to_list()))
    return res


# eval_select ----

SelectExpr: TypeAlias = Union[
    List["str | int"],
    PlSelectExpr,
    str,
    int,
    Callable[[str], bool],
    None,
]
_NamePos: TypeAlias = List[Tuple[str, int]]


@singledispatch
def eval_select(data: DataFrameLike, expr: SelectExpr, strict: bool = True) -> _NamePos:
    """Return a list of column names selected by expr."""

    raise NotImplementedError(f"Unsupported type: {type(expr)}")


@eval_select.register
def _(
    data: PdDataFrame, expr: Union[List[Union[str, int]], Callable[[str], bool]], strict=True
) -> _NamePos:
    if isinstance(expr, str):
        expr = [expr]

    if isinstance(expr, list):
        return _eval_select_from_list(list(data.columns), expr)
    elif callable(expr):
        # TODO: currently, we call on each string, but we could be calling on
        # pd.DataFrame.columns instead (which would let us use pandas .str methods)
        col_pos = {k: ii for ii, k in enumerate(list(data.columns))}
        return [(col, col_pos[col]) for col in data.columns if expr(col)]

    raise NotImplementedError(f"Unsupported selection expr: {expr}")


@eval_select.register
def _(data: PlDataFrame, expr: Union[List[str], _selector_proxy_], strict=True) -> _NamePos:
    # TODO: how to annotate type of a polars selector?
    # Seems to be polars.selectors._selector_proxy_.
    from polars import Expr
    from polars import selectors

    if isinstance(expr, str):
        expr = [expr]

    col_pos = {k: ii for ii, k in enumerate(data.columns)}

    # just in case _selector_proxy_ gets renamed or something
    # it inherits from Expr, so we can just use that in a pinch
    cls_selector = getattr(selectors, "_selector_proxy_", Expr)

    if not isinstance(expr, (list, cls_selector)):
        raise TypeError(f"Unsupported selection expr type: {type(expr)}")

    # I don't think there's a way to get the columns w/o running the selection
    return [(col, col_pos[col]) for col in data.select(expr).columns]


def _eval_select_from_list(columns: list[str], expr: list[str | int]) -> list[tuple[str, int]]:
    col_pos = {k: ii for ii, k in enumerate(columns)}

    # TODO: should prohibit duplicate names in expr?
    res: list[tuple[str, int]] = []
    for col in expr:
        if isinstance(col, str):
            if col in col_pos:
                res.append((col, col_pos[col]))
        elif isinstance(col, int):
            _pos = col if col >= 0 else len(columns) + col
            res.append((columns[col], _pos))
        else:
            raise TypeError(
                f"eval_select received a list with object of type {type(col)}."
                " Only int and str are supported."
            )
    return res


# create_empty ----


@singledispatch
def create_empty_frame(df: DataFrameLike) -> DataFrameLike:
    """Return a DataFrame with the same shape, but all nan string columns"""
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@create_empty_frame.register
def _(df: PdDataFrame):
    import pandas as pd

    return pd.DataFrame(pd.NA, index=df.index, columns=df.columns, dtype="string")


@create_empty_frame.register
def _(df: PlDataFrame):
    import polars as pl

    return df.clear().cast(pl.Utf8).clear(len(df))


@singledispatch
def copy_frame(df: DataFrameLike) -> DataFrameLike:
    """Return a copy of the input DataFrame"""
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@copy_frame.register
def _(df: PdDataFrame):
    return df.copy()


@copy_frame.register
def _(df: PlDataFrame):
    return df.clone()


# cast_frame_to_string ----


@singledispatch
def cast_frame_to_string(df: DataFrameLike) -> DataFrameLike:
    """Return a copy of the input DataFrame with all columns cast to string"""
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@cast_frame_to_string.register
def _(df: PdDataFrame):
    return df.astype("string")


@cast_frame_to_string.register
def _(df: PlDataFrame):
    import polars as pl

    return df.cast(pl.Utf8)


# replace_null_frame ----


@singledispatch
def replace_null_frame(df: DataFrameLike, replacement: DataFrameLike) -> DataFrameLike:
    """Return a copy of the input DataFrame with all null values replaced with replacement"""
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@replace_null_frame.register
def _(df: PdDataFrame, replacement: DataFrameLike):
    return df.fillna(replacement)


@replace_null_frame.register
def _(df: PlDataFrame, replacement: PlDataFrame):
    import polars as pl

    exprs = [pl.col(name).fill_null(replacement[name]) for name in df.columns]
    return df.select(exprs)


@singledispatch
def to_list(ser: SeriesLike) -> List[Any]:
    raise NotImplementedError(f"Unsupported type: {type(ser)}")


@to_list.register
def _(ser: PdSeries) -> List[Any]:
    return ser.tolist()


@to_list.register
def _(ser: PlSeries) -> List[Any]:
    return ser.to_list()


# mutate ----


@singledispatch
def eval_transform(df: DataFrameLike, expr: Any) -> List[Any]:
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@eval_transform.register
def _(df: PdDataFrame, expr: Callable[[PdDataFrame], PdSeries]) -> List[Any]:
    res = expr(df)

    if not isinstance(res, PdSeries):
        raise ValueError(f"Result must be a pandas Series. Received {type(res)}")
    elif not len(res) == len(df):
        raise ValueError(
            f"Result must be same length as input data. Observed different lengths."
            f"\n\nInput data: {len(df)}.\nResult: {len(res)}."
        )

    return res.to_list()


@eval_transform.register
def _(df: PlDataFrame, expr: PlExpr) -> List[Any]:
    df_res = df.select(expr)

    if len(df_res.columns) > 1:
        raise ValueError(f"Result must be a single column. Received {len(df_res.columns)} columns.")
    else:
        res = df_res[df_res.columns[0]]

    if not isinstance(res, PlSeries):
        raise ValueError(f"Result must be a polars Series. Received {type(res)}")
    elif not len(res) == len(df):
        raise ValueError(
            f"Result must be same length as input data. Observed different lengths."
            f"\n\nInput data: {len(df)}.\nResult: {len(res)}."
        )

    return res.to_list()


@singledispatch
def is_na(df: DataFrameLike, x: Any) -> bool:
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@is_na.register
def _(df: PdDataFrame, x: Any) -> bool:
    import pandas as pd

    return pd.isna(x)


@is_na.register
def _(df: PlDataFrame, x: Any) -> bool:
    import polars as pl

    return isinstance(x, (pl.Null, type(None)))


@singledispatch
def validate_frame(df: DataFrameLike) -> DataFrameLike:
    """Raises an error if a DataFrame is not supported by Great Tables.

    Note that this is only relevant for pandas, which allows duplicate names
    on DataFrames, and multi-index columns (and probably other things).
    """
    raise NotImplementedError(f"Unsupported type: {type(df)}")


@validate_frame.register
def _(df: PdDataFrame) -> PdDataFrame:
    import pandas as pd

    # case 1: multi-index columns ----
    if isinstance(df.columns, pd.MultiIndex):
        raise ValueError(
            "pandas DataFrames with MultiIndex columns are not supported."
            " Please use .columns.droplevel() to remove extra column levels,"
            " or combine the levels into a single name per column."
        )

    # case 2: duplicate column names ----
    dupes = df.columns[df.columns.duplicated()]
    if len(dupes):
        raise ValueError(
            f"Column names must be unique. Detected duplicate columns:\n\n {list(dupes)}"
        )

    non_str_cols = [(ii, el) for ii, el in enumerate(df.columns) if not isinstance(el, str)]

    if non_str_cols:
        _col_msg = "\n".join(f"  * Position {ii}: {col}" for ii, col in non_str_cols[:3])
        warnings.warn(
            "pandas DataFrame contains non-string column names. Coercing to strings. "
            "Here are the first few non-string columns:\n\n"
            f"{_col_msg}",
            category=UserWarning,
        )
        new_df = df.copy()
        new_df.columns = [str(el) for el in df.columns]
        return new_df

    return df


@validate_frame.register
def _(df: PlDataFrame) -> PlDataFrame:
    return df
