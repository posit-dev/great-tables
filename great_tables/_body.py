from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
