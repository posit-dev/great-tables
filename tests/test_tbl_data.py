import math
import pandas as pd
import polars as pl
import pyarrow as pa
import polars.testing
import pytest
from great_tables import GT
from great_tables._utils_render_html import create_body_component_h
from great_tables._tbl_data import (
    DataFrameLike,
    SeriesLike,
    _get_cell,
    _get_column_dtype,
    _set_cell,
    _validate_selector_list,
    cast_frame_to_string,
    create_empty_frame,
    eval_select,
    get_column_names,
    group_splits,
    is_series,
    reorder,
    to_frame,
    to_list,
    validate_frame,
    copy_frame,
)

params_frames = [
    pytest.param(pd.DataFrame, id="pandas"),
    pytest.param(pl.DataFrame, id="polars"),
    pytest.param(pa.table, id="arrow"),
]
params_series = [
    pytest.param(pd.Series, id="pandas"),
    pytest.param(pl.Series, id="polars"),
    pytest.param(pa.array, id="arrow"),
    pytest.param(lambda a: pa.chunked_array([a]), id="arrow-chunked"),
]


@pytest.fixture(params=params_frames, scope="function")
def df(request) -> pd.DataFrame:
    return request.param({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})


@pytest.fixture(params=params_series, scope="function")
def ser(request) -> SeriesLike:
    return request.param([1.0, 2.0, None])


def assert_frame_equal(src, target):
    if isinstance(src, pd.DataFrame):
        pd.testing.assert_frame_equal(src, target)
    elif isinstance(src, pl.DataFrame):
        pl.testing.assert_frame_equal(src, target)
    elif isinstance(src, pa.Table):
        assert src.equals(target)
    else:
        raise NotImplementedError(f"Unsupported data type: {type(src)}")


def test_get_column_names(df: DataFrameLike):
    expected = ["col1", "col2", "col3"]
    assert get_column_names(df) == expected


def test_get_column_dtypes(df: DataFrameLike):
    col1 = df["col1"]
    assert _get_column_dtype(df, "col1") == getattr(col1, "dtype", getattr(col1, "type", None))


def test_get_cell(df: DataFrameLike):
    assert _get_cell(df, 1, "col2") == "b"


def test_set_cell(df: DataFrameLike):
    expected_data = {"col1": [1, 2, 3], "col2": ["a", "x", "c"], "col3": [4.0, 5.0, 6.0]}
    if isinstance(df, pa.Table):
        expected = pa.table(expected_data)
    else:
        expected = df.__class__(expected_data)

    new_df = _set_cell(df, 1, "col2", "x")
    if new_df is None:
        # Some implementations do in-place modifications
        new_df = df
    assert_frame_equal(new_df, expected)


def test_reorder(df: DataFrameLike):
    res = reorder(df, [0, 2], ["col2"])

    expected_data = {"col2": ["a", "c"]}
    if isinstance(df, pa.Table):
        dst = pa.table(expected_data)
    else:
        dst = df.__class__(expected_data)

    if isinstance(dst, pd.DataFrame):
        dst.index = pd.Index([0, 2])

    assert_frame_equal(res, dst)


@pytest.mark.parametrize("expr", [["col2", "col1"], [1, 0], ["col2", 0], [1, "col1"]])
def test_eval_select_with_list(df: DataFrameLike, expr):
    sel = eval_select(df, expr)
    assert sel == [("col2", 1), ("col1", 0)]


def test_eval_select_with_callable(df: DataFrameLike):
    def expr(col):
        return col == "col2"

    if isinstance(df, pl.DataFrame):
        # Polars does not support callable expressions
        with pytest.raises(TypeError) as exc_info:
            eval_select(df, expr)
        assert "Unsupported selection expr type:" in str(exc_info.value.args[0])
        return

    sel = eval_select(df, expr)
    assert sel == [("col2", 1)]


@pytest.mark.parametrize(
    "expr",
    [
        pl.selectors.exclude("col3"),
        pl.selectors.starts_with("col1") | pl.selectors.starts_with("col2"),
        [pl.col("col1"), pl.col("col2")],
        [pl.col("col1"), pl.selectors.by_name("col2")],
        pl.col("col1", "col2"),
        pl.all().exclude("col3"),
    ],
)
def test_eval_select_with_list_pl_selector(expr):
    df = pl.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    sel = eval_select(df, expr)
    assert sel == [("col1", 0), ("col2", 1)]


@pytest.mark.parametrize("expr", [["col2", 1.2]])
def test_eval_select_pandas_raises1(expr):
    df = pd.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "Only int and str are supported." in str(exc_info.value.args[0])


@pytest.mark.parametrize("expr", [3.45, {"col2"}, ("col2",)])
def test_eval_select_pandas_raises2(expr):
    df = pd.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(NotImplementedError) as exc_info:
        eval_select(df, expr)

    assert "Unsupported selection expr: " in str(exc_info.value.args[0])


@pytest.mark.parametrize("expr", [3.45, {6}, (7.8,)])
def test_eval_select_polars_raises(expr):
    df = pl.DataFrame({"col1": [1, 2, 3], "col2": ["a", "b", "c"], "col3": [4.0, 5.0, 6.0]})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "Unsupported selection expr type:" in str(exc_info.value.args[0])


def test_eval_selector_polars_list_raises():
    expr = ["col1", 1.2]
    df = pl.DataFrame({"col1": [], "col2": [], "col3": []})
    with pytest.raises(TypeError) as exc_info:
        eval_select(df, expr)

    assert "entry 1 is type: <class 'float'>" in str(exc_info.value.args[0])


@pytest.mark.parametrize("Frame", [pd.DataFrame, pl.DataFrame, pa.table])
def test_group_splits_pd(Frame):
    df = Frame({"g": ["b", "a", "b", "c"]})

    splits = group_splits(df, "g")
    assert set(splits.keys()) == {"a", "b", "c"}
    assert splits["b"] == [0, 2]
    assert splits["a"] == [1]
    assert splits["c"] == [3]


def test_group_splits_pd_na():
    df = pd.DataFrame({"g": ["b", "a", None]})

    splits = group_splits(df, "g")
    assert len(splits.keys()) == 3
    nan_key = [k for k in splits if isinstance(k, float) and math.isnan(k)][0]

    assert splits[nan_key] == [2]


def test_group_splits_pl_na():
    df = pl.DataFrame({"g": ["b", "a", None]})

    splits = group_splits(df, "g")
    assert set(splits.keys()) == {"b", "a", None}
    assert splits[None] == [2]


def test_validate_selector_list_strict_raises():
    with pytest.raises(TypeError) as exc_info:
        _validate_selector_list([pl.col("a")])

    msg = "entry 0 is a polars Expr, which is only supported for polars versions >= 0.20.30."
    assert msg in str(exc_info.value.args[0])


def test_create_empty_frame(df: DataFrameLike):
    res = create_empty_frame(df)
    col = [None] * 3

    if isinstance(res, pd.DataFrame):
        dst = pd.DataFrame({"col1": col, "col2": col, "col3": col}, dtype="string")
    elif isinstance(res, pl.DataFrame):
        dst = pl.DataFrame({"col1": col, "col2": col, "col3": col}).cast(pl.Utf8)
    elif isinstance(res, pa.Table):
        dst = pa.table(
            {"col1": col, "col2": col, "col3": col},
            schema=pa.schema(
                (
                    pa.field("col1", pa.string()),
                    pa.field("col2", pa.string()),
                    pa.field("col3", pa.string()),
                )
            ),
        )
    else:
        raise ValueError(f"Unsupported data type: {type(res)}")

    assert_frame_equal(res, dst)


def test_validate_frame_dupe_cols():
    df = pd.DataFrame([[1, 2, 3]], columns=["x", "x", "y"])

    with pytest.raises(ValueError) as exc_info:
        validate_frame(df)

    assert "Column names must be unique" in str(exc_info.value.args[0])


def test_validate_frame_multi_index():
    df = pd.DataFrame(
        [[1, 2, 3]], columns=pd.MultiIndex.from_tuples([("a", "x"), ("a", "y"), ("b", "x")])
    )

    with pytest.raises(ValueError) as exc_info:
        validate_frame(df)

    assert "MultiIndex columns are not supported" in str(exc_info.value.args[0])


def test_validate_frame_non_str_cols_warning(snapshot):
    df = pd.DataFrame({"x": [1], 55: [2], "y": [3], 99: [4]})

    with pytest.warns(UserWarning) as record:
        validate_frame(df)

    assert len(record) == 1
    assert snapshot == record[0].message.args[0]


def test_validate_frame_non_str_cols_result():
    df = pd.DataFrame({"x": [1], 55: [2], "y": [3], 99: [4]})

    with pytest.warns(UserWarning):
        res = validate_frame(df)

    assert list(res.columns) == ["x", "55", "y", "99"]


def test_to_frame(ser: SeriesLike):
    df = to_frame(ser, name="x")

    if isinstance(ser, pl.Series):
        assert_frame_equal(df, pl.DataFrame({"x": [1.0, 2.0, None]}))
    elif isinstance(ser, pd.Series):
        assert_frame_equal(df, pd.DataFrame({"x": [1.0, 2.0, None]}))
    elif isinstance(ser, (pa.Array, pa.ChunkedArray)):
        assert_frame_equal(df, pa.table({"x": [1.0, 2.0, None]}))
    else:
        raise AssertionError(f"Unexpected series type: {type(ser)}")


def test_is_series(ser: SeriesLike):
    assert is_series(ser)


def test_is_series_false():
    assert not is_series(1)


def test_to_list(ser: SeriesLike):
    pylist = to_list(ser)
    assert len(pylist) == 3
    assert pylist[:2] == [1.0, 2.0]


def test_cast_frame_to_string_polars_list_col():
    df = pl.DataFrame({"x": [[1, 2], [3]], "y": [1, None], "z": [{"a": 1}, {"a": 2}]})
    new_df = cast_frame_to_string(df)

    assert new_df["x"].dtype.is_(pl.String)
    assert new_df["y"].dtype.is_(pl.String)
    assert new_df["z"].dtype.is_(pl.String)


def test_frame_rendering(df: DataFrameLike, snapshot):
    gt = GT(df).fmt_number(columns="col3", decimals=0).fmt_currency(columns="col1")
    assert create_body_component_h(gt._build_data("html")) == snapshot


def test_copy_frame(df: DataFrameLike):
    copy_df = copy_frame(df)
    assert id(copy_df) != id(df)
    assert_frame_equal(copy_df, df)
