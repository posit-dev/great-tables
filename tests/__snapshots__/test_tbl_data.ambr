# serializer version: 1
# name: test_frame_rendering[arrow]
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">$1.00</td>
      <td class="gt_row gt_left">a</td>
      <td class="gt_row gt_center">4</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$2.00</td>
      <td class="gt_row gt_left">b</td>
      <td class="gt_row gt_center">5</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$3.00</td>
      <td class="gt_row gt_left">c</td>
      <td class="gt_row gt_center">6</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_frame_rendering[pandas]
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">$1.00</td>
      <td class="gt_row gt_left">a</td>
      <td class="gt_row gt_right">4</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$2.00</td>
      <td class="gt_row gt_left">b</td>
      <td class="gt_row gt_right">5</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$3.00</td>
      <td class="gt_row gt_left">c</td>
      <td class="gt_row gt_right">6</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_frame_rendering[polars]
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">$1.00</td>
      <td class="gt_row gt_left">a</td>
      <td class="gt_row gt_right">4</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$2.00</td>
      <td class="gt_row gt_left">b</td>
      <td class="gt_row gt_right">5</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">$3.00</td>
      <td class="gt_row gt_left">c</td>
      <td class="gt_row gt_right">6</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_validate_frame_non_str_cols_warning
  '''
  pandas DataFrame contains non-string column names. Coercing to strings. Here are the first few non-string columns:
  
    * Position 1: 55
    * Position 3: 99
  '''
# ---
